### PR TITLE
[patch] change gateway vald's mutex lock

### PR DESCRIPTION
**Pull request by kpango**
Signed-off-by: kpango <i.can.feel.gravity@gmail.com>

### Description:
replace sync.Mutex lock map using sync.Map CAS operation

### Environment:

- Go Version: 1.15.2
- Docker Version: 19.03.8
- Kubernetes Version: 1.18.2
- NGT Version: 1.12.1

### Types of changes:

- [ ] Bug fix [type/bug]
- [ ] New feature [type/feature]
- [ ] Add tests [type/test]
- [ ] Security related changes [type/security]
- [ ] Add documents [type/documentation]
- [ ] Refactoring [type/refactoring]
- [ ] Update dependencies [type/dependency]
- [ ] Update benchmarks and performances [type/bench]
- [ ] Update CI [type/ci]

### Changes to Core Features:

- [ ] Have you added an explanation of what your changes do and why you'd like us to include them?
- [ ] Have you written new tests for your core changes, as applicable?
- [ ] Have you successfully ran tests with your changes locally?

### Checklist:

- [ ] I have read the [CONTRIBUTING](https://github.com/vdaas/vald/blob/master/CONTRIBUTING.md) document.
- [ ] I have checked open [Pull Requests](https://github.com/vdaas/vald/pulls) for the similar feature or fixes?
- [ ] I have added tests and benchmarks to cover my changes.
- [ ] I have ensured all new and existing tests passed.
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have updated the documentation accordingly.

### CODE REVIEW

File: pkg/gateway/vald/handler/grpc/handler.go [hunk 1]
```diff
@@ -217,44 +209,44 @@ func (s *server) search(ctx context.Context, cfg *payload.Search_Config,
 			}
 			return res, nil
 		case dist := <-dch:
-			if len(res.GetResults()) >= num {
-				if dist.GetDistance() < math.Float32frombits(atomic.LoadUint32(&maxDist)) {
-					atomic.StoreUint32(&maxDist, math.Float32bits(dist.GetDistance()))
-				} else {
-					continue
-				}
+			pos := len(res.GetResults())
```
Comment (rinx) [inline pkg/gateway/vald/handler/grpc/handler.go:212]: `pos` is misleading name. please rename it like `resultLen`, `resLen` or something.

File: pkg/gateway/vald/handler/grpc/handler.go [hunk 2]
```diff
@@ -217,44 +209,44 @@ func (s *server) search(ctx context.Context, cfg *payload.Search_Config,
...
-				continue
-			}
-
-			pos := len(res.GetResults())
-			for idx := pos; idx >= 1; idx-- {
-				if res.GetResults()[idx-1].GetDistance() <= dist.GetDistance() {
-					pos = idx - 1
-					break
+			default:
+				for idx := pos; idx >= 1; idx-- {
```
Comment (rinx) [inline pkg/gateway/vald/handler/grpc/handler.go:226]: using `resLen`,
```suggestion
				pos := resLen
				for idx := pos; idx >= 1; idx-- {
```

File: pkg/gateway/vald/handler/grpc/handler.go [hunk 3]
```diff
@@ -217,44 +209,44 @@ func (s *server) search(ctx context.Context, cfg *payload.Search_Config,
...
+				switch {
+				case pos == len(res.GetResults()):
+					res.Results = append([]*payload.Object_Distance{dist}, res.Results...)
+				case pos == len(res.GetResults())-1:
+					res.Results = append(res.GetResults(), dist)
+				case pos >= 0:
+					res.Results = append(res.GetResults()[:pos+1], res.GetResults()[pos:]...)
+					res.Results[pos+1] = dist
+				}
+				pos = len(res.GetResults())
```
Comment (rinx) [inline pkg/gateway/vald/handler/grpc/handler.go:241]: also, `pos` is misleading name here.

**Review by rinx — CHANGES_REQUESTED** [review]

File: pkg/gateway/vald/handler/grpc/handler.go (hunk 1 above)
Comment (kpango) [reply pkg/gateway/vald/handler/grpc/handler.go:212]: position is misreading??

Comment (rinx) [reply pkg/gateway/vald/handler/grpc/handler.go:212]: what is the position of? here, this variable is just used for L213 and L216.
they should be treated as a different variable from the variable on L226 (that is the "true" `pos` variable).

Comment (kpango) [reply pkg/gateway/vald/handler/grpc/handler.go:212]: position of result end index

Comment (rinx) [reply pkg/gateway/vald/handler/grpc/handler.go:212]: you should name it as like that, i think.

Comment (kpango) [reply pkg/gateway/vald/handler/grpc/handler.go:212]: hmm, anything is okay, it's not important

Comment (kpango) [reply pkg/gateway/vald/handler/grpc/handler.go:212]: I'll change it and copy it to pos

File: pkg/gateway/vald/handler/grpc/handler.go [hunk 4]
```diff
@@ -130,7 +130,8 @@ func (s *server) search(ctx context.Context, cfg *payload.Search_Config,
 			span.End()
 		}
 	}()
-	maxDist := uint32(math.MaxUint32)
+	var maxDist uint32
+	atomic.StoreUint32(&maxDist, math.Float32bits(math.MaxFloat32))
```
Comment (kmrmt) [inline pkg/gateway/vald/handler/grpc/handler.go:134]: I think `float32(math.Inf(0))` is more precise representation than `math.MaxFloat32`, but it makes little difference in practical.
How about it?

Comment (kpango) [reply pkg/gateway/vald/handler/grpc/handler.go:134]: math.Inf returns float64 we can't use here

**Comment by kpango** [comment]
@rinx @kmrmt fixed,

File: pkg/gateway/vald/handler/grpc/handler.go [hunk 5]
```diff
@@ -217,44 +209,46 @@ func (s *server) search(ctx context.Context, cfg *payload.Search_Config,
 			}
 			return res, nil
 		case dist := <-dch:
-			if len(res.GetResults()) >= num {
-				if dist.GetDistance() < math.Float32frombits(atomic.LoadUint32(&maxDist)) {
-					atomic.StoreUint32(&maxDist, math.Float32bits(dist.GetDistance()))
-				} else {
-					continue
-				}
+			rl := len(res.GetResults())
```
Comment (rinx) [inline pkg/gateway/vald/handler/grpc/handler.go:212]: please follow the coding style, https://github.com/vdaas/vald/blob/master/docs/contributing/coding-style.md#variables-and-constant
```suggestion
			rl := len(res.GetResults()) // result length
```

**Comment by kpango** [comment]
@rinx commit suggestion

**Comment by rinx** [comment]
Thanks! it became understandable code. :+1:

**Comment by kpango** [comment]
kmrmr's review we don't need to fix it. I'll merge this PR

**Pull request merged**